### PR TITLE
Add reproducible build spec

### PR DESCRIPTION
**Pull request by clairernovotny**
This spec covers the reproducible builds the .NET, NuGet, and Terrapin teams have been working on.

### CODE REVIEW

**Comment by clairernovotny** [comment]
@tmat, @RikkiGibson, @clairernovotny, @ryzngard, @terrajobst, @richlander

**Comment by clairernovotny** [comment]
The implementation spec is here 
https://github.com/dotnet/roslyn/pull/49886

File: accepted/2020/reproducible-builds.md [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+**Future**
+
+- Support for IL weavers
+
+### Non-Goals
+
+- Evaluating provenance of embedded resources, such as .resx or .baml.
+- Support for modification or patching during validation, to create a runnable binary that's functionally different from the original.
+- Support for exporting a rebuilt artifact
```
Comment (omajid) [inline accepted/2020/reproducible-builds.md:100]: Doesn't this seem counter to this part in the Overviews? 

> Also, should package consumers find a zero-day exploit, they usually have no way to patch the binaries themselves. They must find the corresponding repo, get the necessary dependencies for building it, somehow find the correct version of the code, patch it, and then rebuild. Not an easy task under ideal circumstances, much less viable when time is of the essence.

If the tool doesn't produce (and reproduce) the original environment and the rebuilt artifact, this part of the user story is still not addressed.

File: accepted/2020/reproducible-builds.md [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+- Supporting the tool being used to validate binaries in an environment different than what they were produced on.
```
Comment (omajid) [inline accepted/2020/reproducible-builds.md:101]: What does "environment" here mean? Is it just the architecture + OS combination? Or is there more to it?

File: accepted/2020/reproducible-builds.md (hunk 1 above)
Comment (clairernovotny) [reply accepted/2020/reproducible-builds.md:100]: The explicit goal for v1 is to enable provenance traceability between the binary and the source. In order to do that, the dll will be rebuilt and the result compared, however using the toolchain for patching isn't an explicit goal initially.

Other issues like source generators arise where the source generator won't be re-run here. Depending on what the source generator does, making a change to the source may or may not work.

File: accepted/2020/reproducible-builds.md [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+
+A developer is updating their library and wants ensure that it can be successfully validated so their users feel secure in taking a dependency on it, and that NuGet.org will give it a higher health score.
+
+The developer knows that by updating to the latest .NET SDK, they are secure by default because the right settings are automatically enabled for their project:
+
+- Source Link
+- EmbedUntrackedSources
+- PublishRepositoryUrl
+- Latest Roslyn compiler which stores compiler flags into the PDB
+- PDB is automatically included in the NuGet package
```
Comment (bruno-garcia) [inline accepted/2020/reproducible-builds.md:47]: Is this contrary to the last guidelines that say we should use symbol packages?

Asking to know [if we need to revert this.](https://github.com/getsentry/sentry-dotnet/pull/552)

File: accepted/2020/reproducible-builds.md [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+
+- Package consumers who wish to manually check results on their own.
+- Package authors who wish to verify their package is reproducible.
+- Package hosts, such as NuGet.org and Terrapin, who wish to validate packages on upload and present the status as part of overall package/project health.
+
+### End-to-end scenarios
+
+#### Developer or InfoSec that wants to validate on their own
+
+An internal package security team is wants to reduce their supply chain risk to their apps and services. They want to make sure that the binaries being used can be traced to sources so they can run security scans on the source code. The packages being validated are produced from a mix of public and internal sources. The team sets up infrastructure to run the validation tool for each package they use.
```
Comment (ryzngard) [inline accepted/2020/reproducible-builds.md:30]: Very nit grammar 

```suggestion
An internal package security team wants to reduce their supply chain risk to their apps and services. They want to make sure that the binaries being used can be traced to sources so they can run security scans on the source code. The packages being validated are produced from a mix of public and internal sources. The team sets up infrastructure to run the validation tool for each package they use.
```

File: accepted/2020/reproducible-builds.md [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+
+To address these problems, we will create a set of tooling that for each managed binary file in a NuGet package, will be able to enumerate all of the input sources used to create it, validate they are accessible and validate that the resulting binary matches was produced using those sources files. In doing so, we don't need to trust the original build environment, as we're able to independently create the managed file from the input sources.
+
+## Scenarios and User Experience
+
+### Customer Cohorts
+
+- Package consumers who wish to manually check results on their own.
+- Package authors who wish to verify their package is reproducible.
+- Package hosts, such as NuGet.org and Terrapin, who wish to validate packages on upload and present the status as part of overall package/project health.
```
Comment (ryzngard) [inline accepted/2020/reproducible-builds.md:24]: One cohort not mentioned here, but was mentioned in introduction, is cohort who wants to patch a binary.

File: accepted/2020/reproducible-builds.md [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+
+### End-to-end scenarios
+
+#### Developer or InfoSec that wants to validate on their own
+
+An internal package security team is wants to reduce their supply chain risk to their apps and services. They want to make sure that the binaries being used can be traced to sources so they can run security scans on the source code. The packages being validated are produced from a mix of public and internal sources. The team sets up infrastructure to run the validation tool for each package they use.
+
+Due to organization security requirements, based on likely forthcoming NIST standards, they must be able to trace the source and rebuild all packages being used.
+
+1. For each managed binary in the NuGet package, call a tool to check the status. Result is one of: not buildable, buildable but not verifiable (builds but not the same deterministic output), or deterministically reproducible.
```
Comment (ryzngard) [inline accepted/2020/reproducible-builds.md:34]: "buildable but not verifiable" : Is there a reason for this specific wording? It seems to indicate that tool could build but may have failed in some way to do a verification step. Maybe "buildable but not verified". Could even go for more severe wording like "built but failed verification"

File: accepted/2020/reproducible-builds.md (hunk 3 above)
Comment (ryzngard) [reply accepted/2020/reproducible-builds.md:47]: If snupkg contains portable pdb I would think it's still a viable strategy for this. At the end of the day we just need a correlation from binary to portable pdb.

File: accepted/2020/reproducible-builds.md [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+## Requirements
+
+### Goals
+
+**P0**
+
+- The validation tool must work on any managed binary, regardless of target framework
+- The validation tool must work on any operating system.
+- The validation tool must support public sources for sources and symbols.
+- The validation tool must be easy to use: single command the user invokes.
```
Comment (ryzngard) [inline accepted/2020/reproducible-builds.md:81]: I would add goals about diagnostic/investigative targets for the tool. Users will need to be able to follow up problems they run into and understand next steps they should take.

File: accepted/2020/reproducible-builds.md (hunk 2 above)
Comment (ryzngard) [reply accepted/2020/reproducible-builds.md:101]: Ideally we support all inputs which result in a deterministic build. https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/deterministic-compiler-option

File: accepted/2020/reproducible-builds.md [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+
+The tools require the following artifacts to work:
+
+- Portable PDBs (embedded is supported)
+- Binary references (referenced assemblies)
+- Source
+
+### PDBs
+
+PDBs are a critical component as they contain pointers to the original source, or embeded source. They also contain information about the binary references and compiler flags used during the original compilation.
```
Comment (ryzngard) [inline accepted/2020/reproducible-builds.md:123]: We should note the compiler version that is required for production of PDB here, and that they have to be portable PDB.

File: accepted/2020/reproducible-builds.md [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+Downloadng reference files c:\temp\references
+... trying to get System.dll from msdl.microsoft.com: Found
+... trying to get AutoMapper.dll from msdl.microsoft.com: Not Found
+... trying to get AutoMapper.dll from symbols.nuget.org: Found
+All artifacts downloaded, validating...
+
+File validated successfully.
+```
+
+Error conditions will output relevant troublshooting data:
```
Comment (ryzngard) [inline accepted/2020/reproducible-builds.md:158]: Love this

File: accepted/2020/reproducible-builds.md (hunk 3 above)
Comment (bruno-garcia) [reply accepted/2020/reproducible-builds.md:47]: Got it.
What is the strategy for correlating those? Mono uses mvid (module version id), on mono-symbolicate. Is this how we're doing things here too?

Comment (tmat) [reply accepted/2020/reproducible-builds.md:47]: Yes, both including PDB in nupkg or snupkg works. Here's the spec of CodeView Debug Directory entrye, which ties binary to the PDB: https://github.com/dotnet/runtime/blob/master/docs/design/specs/PE-COFF.md#codeview-debug-directory-entry-type-2

File: accepted/2020/reproducible-builds.md (hunk 6 above)
Comment (jaredpar) [reply accepted/2020/reproducible-builds.md:34]: In general the verb we were using for the provenance step was "verify" hence "verifiable" makes sense in that context. 

Personally I'm not wedded to "verify" as the verb here but we also couldn't find another one that read well.

File: accepted/2020/reproducible-builds.md (hunk 5 above)
Comment (jaredpar) [reply accepted/2020/reproducible-builds.md:24]: In more recent discussions that cohort was essentially moved out of the primary scenarios. The focus at this point was more on provenance validation vs. patching. I agree though that's a scenario, in addition to several others, that we will likely expand into in the future.

**Comment by clairernovotny** [comment]
Merging. As we move forward, we'll revise as appropriate.

**Pull request merged**